### PR TITLE
dump: fix dump/config race conditions, preserve message order, add unit tests

### PR DESCRIPTION
**Pull request by klihub**
There are a number of race conditions in pkg/dump. The most prominent one is related to concurrent message dumping and dynamic reconfiguration. This one is relatively easy to trigger on  a busy node. It also gets immediately triggered when one starts adding other than trivial end-to-end tests.

This PR reworks pkg/dump to fix these race conditions and a number of its other problems. In particular the PR
  - reworks message dumping to be race-free
  - preserves the order of messages
  - simplifies configuration by dropping command line support
  - adds unit tests

### CODE REVIEW

File: pkg/dump/dump.go [hunk 1]
```diff
@@ -45,175 +46,303 @@ const (
...
+		d.methods[idx] = method
+		d.details[method] = detail
+	}
+}
+
+// name does a name-only dump of the given message.
+func (d *dumper) name(dir direction, kind, method string, msg interface{}, latency time.Duration) {
+	switch dir {
+	case request:
+		return
```
Comment (marquiz) [inline pkg/dump/dump.go:238]: We don't dump requests?

File: pkg/dump/dump.go [hunk 2]
```diff
@@ -45,175 +46,303 @@ const (
...
-func dumpToFile(dir string, latency time.Duration, format string, args ...interface{}) {
-	if !checkDumpFile() {
+// tofile dumps a single line to a file.
+func (d *dumper) tofile(dir direction, latency time.Duration, format string, args ...interface{}) {
+	if d.file == nil {
 		return
 	}
-
-	fmt.Fprintf(file, "["+stamp(dir, latency)+"] "+format+"\n", args...)
+	fmt.Fprintf(d.file, "["+stamp(dir, latency)+"] "+format+"\n", args...)
```
Comment (marquiz) [inline pkg/dump/dump.go:315]: nit: minuscule simplification
```
	if d.file != nil {
		fmt.Fprintf(d.file, "["+stamp(dir, latency)+"] "+format+"\n", args...)
	}
```

File: pkg/dump/dump.go (hunk 1 above)
Comment (klihub) [reply pkg/dump/dump.go:238]: With name-only dumping we dump request/reply-pairs as a single line, and only once the reply has come in, so we know whether the request succeeded or failed.

File: pkg/dump/dump.go [hunk 3]
```diff
@@ -45,175 +46,303 @@ const (
...
+	defer dump.Unlock()
+	dump.train(methods)
+}
+
+// RequestMessage dumps a CRI request.
+func RequestMessage(kind, name string, req interface{}, sync bool) {
+	if !dump.disabled {
+		var ch chan struct{}
+		if sync {
+			ch = make(chan struct{})
```
[the remark this answers is not in the document]
Comment (klihub) [reply pkg/dump/dump.go:100]: No, this channel is only used for signalling. I should fix the other instances which apparently I forgot in a rush to change to get correctly created with 0 length.

**Pull request merged**